src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -267,14 +344,20 @@ struct FindComponentNearestNeighbors
       }
     } while (node != SENTINEL);
 
-    // This check is only here to reduce hammering the atomics for large
-    // components. Otherwise, for a large number of points and a small number of
-    // components it becomes extremely expensive.
-    auto &component_edge = _edges(component - n + 1);
-    if (current_best.weight < inf &&
-        current_best.weight <= component_edge.weight)
+    // Using atomics on WeightedEdge is exteremely slow when they are called
+    // often (e.g., PortoTaxi problem). Instead, we work around this by using
+    // several parallel loops to get the same result. Most of the loops are
+    // inside atomicComponentEdgesWorkaround, here we set the component edge
+    // weights.
+    if (current_best.weight < inf)
     {
-      Kokkos::atomic_min(&component_edge, current_best);
+      auto &component_edge = _edges(component - n + 1);
+      Kokkos::atomic_min(&component_edge.weight, current_best.weight);
+
+      // We want to reduce the number of memory writes as much as possible, so
+      // only do it when absolutely necessary.
+      if (current_best.weight == component_edge.weight)
+        _found_edges(i - n + 1) = current_best;
     }
```

Comment:
Note that I actually tried doing a simple unconditional
```c++
    _found_edges(i - n + 1) = current_best;
```
but that slowed down HACC iterations by 30%+. So this patch minimizes the number of writes to `_found_edges`.